keps/582-preempt-based-on-flavor-order/README.md

```diff
@@ -0,0 +1,419 @@
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Cluster Queue API
+
+We extend the Cluster Queue API to introduce the new fields: flavorFungibility to opt-in and configure the new behavior.
+
+For each type of resource in each podSet, Kueue will traverse all resource groups and resource flavors to find a available flavor in present. When there are insufficient resources in the flavor, kueue will prioritize preemption or borrowing based on the configured policy. 
+
+```
+const (
+	Borrow FlavorFungibilityPolicy = "Borrow"
+	Preempt  FlavorFungibilityPolicy = "Preempt"
+  TryNextFlavor FlavorFungibilityPolicy = "TryNextFlavor"
+)
+
+type FlavorFungibility struct {
+  // +kubebuilder:validation:Enum="Borrow,TryNextFlavor"
+  WhenCanBorrow FlavorFungibilityPolicy  `json:"whenCanBorrow"`
+  // +kubebuilder:validation:Enum="Preempt,TryNextFlavor"
+  WhenCanPreempt FlavorFungibilityPolicy `json:"whenCanPreempt"`
+}
+
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+	...
+	FlavorFungibility FlavorFungibility `json:"flavorFungibility"`
+}
+```
+
+If flavorFungibility is nil in configuration, we will set the `WhenCanBorrow` to `true` and set `WhenCanPreempt` to `false` to maintain consistency with the current behavior.
+
+### Behavior Changes
+
+We will not change the behavior to judge whether a podset can get enough resource in certain resource flavor. Preemption and admission will not be influenced also. We only change the order these flavors were considered.
+
+After we try to schedule a podset in a resource flavor, we decide whether to traverse to the next flavor base on the `flavorFungibility` like the follows. This turn run to end if `shouldTryNextFlavor` return false, otherwise we try next flavor to find the best one. 
+```
+func shouldTryNextFlavor(representativeMode FlavorAssignmentMode, flavorFungibility v1beta1.FlavorFungibility, whetherNeedBorrowing bool) bool {
+	policyPreempt := flavorFungibility.WhenCanPreempt
+	policyBorrow := flavorFungibility.WhenCanBorrow
+	if representativeMode == Preempt && policyPreempt == v1beta1.Preempt {
+		return false
+	}
+
+	if representativeMode == Fit && whetherNeedBorrowing && policyBorrow == v1beta1.Borrow {
+		return false
+	}
+
+	if representativeMode == Fit && !whetherNeedBorrowing {
+		return false
+	}
+
+	return true
+}
+```
+
+We will store the scheduling context in workload info so that we can start from where we stop in previous scheduling attempts. This will be useful to avoid to waste time in one flavor all the time if we try to preempt in a flavor and failed. Scheduling context will contain the `LastScheduledFlavorIdx`, `ResourceFlavors` attached to the CQ, `ClusterQueueUsage` of the CQ and `CohortUsage`. Any changes to these properties will lead to a scheduling from the first flavor.
```

Comment:
Some implementation detail: if the flavors in the CQ change, we should wipe the stored context. Maybe we can keep a hash to detect the changes.